### ISSUE TITLE
Scylla should flush stale memtables periodically.

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#15920]
As discussed in https://github.com/scylladb/scylladb/pull/15820#issuecomment-1790274807,
after https://github.com/scylladb/scylladb/pull/15060, compaction tombstone garbage-collection may be delayed indefinitely while commitlog segments may hold data shadowed by expired tombstones.

Unrelated tables that are written infrequently may "pin" down some commitlog segments indefinitely, until their memtables are flushed.  Since there is no guarantee that they will ever be flushed, we should arm a timer so that if a non-empty memtable is not flushed within the idle timeout period (e.g. 1 hour by default), it will be auto-flushed by the timer.
This will provide a good trade-off between increased write-amplification due to excessive flushing (if this timeout was set to 1 second for example), vs. space- and read-amplification due to holding-off tombstone garbage-collection.

I suggest that the said timeout default value would be configurable.
If needed, per-table tuning may later be added by either extending the table schema, or by adding a more detailed configuration option in scylla.yaml

Cc @avikivity @tzach @vladzcloudius @raphaelsc @elcallio

**Comment by bhalevy** [comment]
@denesb I think this falls nicely in your team's realm of responsibility

**Comment by elcallio** [comment]
Should we not try to observe this behaviour in nature before reacting to it? In any real scenario, even if X tables are low-volume, there should be some write activity -> eventually CL fills up and reaches usage threshold. Then all tables occupying non-clean segments will be flushed, and said tombstones will become eligible.

While we could potentially see GC delays with pathologically short TTL:s for tombstones and a no-throughput workload, is that really a real scenario?

**Comment by raphaelsc** [comment]
> Should we not try to observe this behaviour in nature before reacting to it? In any real scenario, even if X tables are low-volume, there should be some write activity -> eventually CL fills up and reaches usage threshold. Then all tables occupying non-clean segments will be flushed, and said tombstones will become eligible.
> 

So we flush all tables in an active segment that hit usage threshold? And given that tables co-habit segments, their write frequency is not a problem given that a table with high write frequency will cause even almost-idle tables to have their memtables flushed from time to time?

**Comment by elcallio** [comment]
Yes. Once data usage in CL hits a certain level, all tables that hog segments will be requested to flush. 
Obviously though, there are pathological/theoretical scenarios here where segments could remain ignored and hog things almost indefinitely. (a non-active table has a mutation in a segment that blocks GC, but all other segments are released without CL having to force flush, so never reach threshold -> hog forever). But I doubt that will actually happen. 

I am not _against_ having an unconditional, time-based flush factor as well here, I am just saying the case of really needing it should be consider (semi-)artificial/pathologic, so should try to avoid intrusiveness.

**Comment by bhalevy** [comment]
> > Yes. Once data usage in CL hits a certain level, all tables that hog segments will be requested to flush. Obviously though, there are pathological/theoretical scenarios here where segments could remain ignored and hog things almost indefinitely. (a non-active table has a mutation in a segment that blocks GC, but all other segments are released without CL having to force flush, so never reach threshold -> hog forever). But I doubt that will actually happen.
> > I am not _against_ having an unconditional, time-based flush factor as well here, I am just saying the case of really needing it should be consider (semi-)artificial/pathologic, so should try to avoid intrusiveness.
> 
> I thought the whole discussion started from a pathological case from a customer.

Right

**Comment by elcallio** [comment]
The original discussion was on the inverse; data resurrection. Now we're instead discussing how preventing data resurrection instead can lead to GC starvation.
